staging/src/k8s.io/client-go/tools/portforward/portforward.go

```diff
@@ -37,7 +38,22 @@ import (
 // TODO move to API machinery and re-unify with kubelet/server/portfoward
 const PortForwardProtocolV1Name = "portforward.k8s.io"
 
-var ErrLostConnectionToPod = errors.New("lost connection to pod")
+var (
+	// error returned whenever we lost connection to a pod
+	ErrLostConnectionToPod = errors.New("lost connection to pod")
+
+	// set of error we're expecting during port-forwarding
+	networkClosedError = "use of closed network connection"
+	// Containerd will return "network namespace for sandbox %q is closed", see:
+	// - 2.0: https://github.com/containerd/containerd/blob/v2.0.0-rc.6/internal/cri/server/sandbox_portforward_linux.go#L47
+	// - 1.7: https://github.com/containerd/containerd/blob/release/1.7/pkg/cri/server/sandbox_portforward_linux.go#L47
+	// CRI-O will return "sandbox %s is not running", see:
+	// - 1.31: https://github.com/cri-o/cri-o/blob/release-1.31/server/container_portforward.go#L54
+	networkClosedRe = regexp.MustCompile(`network namespace for sandbox "\w+" is closed|sandbox \w+ is not running`)
+	// this error matches the one returned in this file, in handleConnection, when we create
+	// data stream - https://github.com/kubernetes/kubernetes/blob/release-1.31/staging/src/k8s.io/client-go/tools/portforward/portforward.go#L374
```

Comment:
https://github.com/kubernetes/kubernetes/blob/release-1.31/staging/src/k8s.io/client-go/tools/portforward/portforward.go#L374 doesn't return that error message, it just logs it... I think that means our detection of failure to *create* a stream will never fire, and we are 100% relying on detecting specific strings from the backend to avoid a permanently failing portforward loop?

Reply:
If you dropped the `networkClosedRe` condition, I would expect portforward on a dead connection to close on the next attempt, but I don't think that would work properly. Do we have a test that exercises that?

Reply:
yeah, I think this is the point we discussed last time, to not depend on the errors coming down from the error stream to detect if the problem is terminal.
If we reset() the stream we no longer need to do matching on the error stream that is coming down the pipe to fix the "soft errors"... I thought that for knowing if the error was terminal we were going to implement client side a mechanism to check if a new connection after the error wil be able to proceed, and if not,  we completely shutdown  the listener

Reply:
> If you dropped the networkClosedRe condition, I would expect portforward on a dead connection to close on the next attempt, but I don't think that would work properly. 

It would not close, because we only close the whole connection when matching that error. Which happens in 2 cases:
1. the network closed error, which happens when a pod is removed
2. with older CRI (which doesn't reset connection) you'll get the stream creation timeout

> Do we have a test that exercises that?

Not directly, because this was strictly to cover the case of an older CRI (ie. w/o the reset fix), where subsequent stream creation fails. Lemme double check where that error is coming from again, and if both crio and containerd return just that.

Reply:
I did some more testing, and following your advice I dropped the networking checks, especially that after I started testing I realized that I only addressed the problem when you're working in a used port-forward, but if the pod disappears in a different point (for example before even opening the port-forward, you'll actually get a different error [this for containerd](https://github.com/containerd/containerd/blob/release/1.7/pkg/cri/server/sandbox_portforward_linux.go#L35) and [this for crio](https://github.com/cri-o/cri-o/blob/release-1.31/server/container_portforward.go#L45)). So at the end of the day, we'd need to account for all possible errors, which would only unnecessarily complicate this. 

So in the last commit, I dropped the network checks (including swapping the `io.Copy` check that Antonio mentioned). I've added a logic which handles the stream timeout, which then closes the connection. The downside of that approach is that:
1. we have a direct spdystream dependency in client-go, which if we decide to drop we can replace with a string match.
2. one of the tests (the one testing continuity of the connection) won't pass so I had to comment it out, until we propagate the fix from #128318 into crio and containerd, at minimum.

Opinions?

Reply:
> 1. we have a direct spdystream dependency in client-go, which if we decide to drop we can replace with a string match.

A better approach could be exposing the errorcheck inside the `k8s.io/apimachinery/pkg/util/httpstream` package, which hides the spdy|websocket implementation details.